src/Control/Monad/Bayes/Traced/Common.hs

```diff
@@ -65,21 +65,32 @@ bind dx f = do
   t2 <- f (output t1)
   return $ t2 {variables = variables t1 ++ variables t2, density = density t1 * density t2}
 
--- | A single Metropolis-corrected transition of single-site Trace MCMC.
-mhTrans :: MonadSample m => Weighted (FreeSampler m) a -> Trace a -> m (Trace a)
-mhTrans m t@Trace {variables = us, density = p} = do
+-- | a proposal distribution over randomness
+singleVariableProposal :: MonadSample m => [Double] -> m [Double]
+singleVariableProposal us = do
+  let n = length us
+  i <- discrete $ discreteUniformAB 0 (n -1)
+  u' <- random
+  case splitAt i us of
+    (xs, _ : ys) -> return $ xs ++ (u' : ys)
+    _ -> error "impossible"
+
+-- | A single Metropolis-corrected transition of single-site Trace MCMC
+-- | with a proposal distribution p([x_j]^i+1 | [x_j]^i). The x_j correspond
+-- | to the randomness of the sample, defined over the domain [0, 1].
```

Comment:
What's `i`?

Reply:
the same as in the original code: the index of the random variable that get's modified in the proposal distribution to another sample from `uniform 0 1`. All other random variables are kept fix.